File: 20. Valid Parentheses.rb
def is_valid(s)
  stack = []
  parentheses = {
    "(" => ")",
    "[" => "]",
    "{" => "}",
  }
  
  # check first element
  if (s[0] === ")" || s[0] === "}" || s[0] === "]")
    return false
  end

  if (s.length.odd?)
    return false
  end
  
  for i in 0..(s.length-1) do
    stack.push(s[i])
    if (stack[stack.length-1] === ")" ||
        stack[stack.length-1] === "}" ||
        stack[stack.length-1] === "]")
      if (parentheses[stack[stack.length-2]] === stack[stack.length-1])
        stack.pop()
        stack.pop()
      else
        return false
      end
    end
  end

  if (stack.length > 0)
    return false
  end
  
  return true
end

File: 20. Valid Parentheses.ts
function isValid(s: string): boolean {
    const parentheses = {
        "[": "]",
        "{": "}",
        "(": ")",
    };
    const stack: string[] = [];

    for (let i = 0; i < s.length; i++) {
        if (s[i] in parentheses) stack.push(s[i]);
        else {
            const openBracket = stack[stack.length-1];
            if (parentheses[openBracket] !== s[i]) return false;
            stack.pop();
        }
    }
    
    return !stack.length;
};

// 0(n)

